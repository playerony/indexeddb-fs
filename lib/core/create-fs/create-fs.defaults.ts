import { CreateFsProps } from './create-fs.types';

export const defaultProps: Required<CreateFsProps> = {
  databaseVersion: 1,
  objectStoreName: 'files',
  rootDirectoryName: 'root',
  databaseName: 'indexeddb-fs',
};
