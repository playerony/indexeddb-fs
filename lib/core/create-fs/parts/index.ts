export * from './exists-decorator/exists-decorator.function';
export * from './remove-decorator/remove-decorator.function';
export * from './is-file-decorator/is-file-decorator.function';
export * from './read-file-decorator/read-file-decorator.function';
export * from './write-file-decorator/write-file-decorator.function';
export * from './remove-file-decorator/remove-file-decorator.function';
export * from './is-directory-decorator/is-directory-decorator.function';
export * from './read-directory-decorator/read-directory-decorator.function';
export * from './create-directory-decorator/create-directory-decorator.function';
export * from './remove-directory-decorator/remove-directory-decorator.function';
