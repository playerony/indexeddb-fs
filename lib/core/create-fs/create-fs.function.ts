import {
  existsDecorator,
  isFileDecorator,
  removeDecorator,
  readFileDecorator,
  writeFileDecorator,
  removeFileDecorator,
  isDirectoryDecorator,
  readDirectoryDecorator,
  createDirectoryDecorator,
  removeDirectoryDecorator,
} from './parts';
import {
  isIndexedDBSupport,
  validateCreateFSProps,
  initializeObjectStoreDecorator,
} from '@core/utils';

import { CreateFSProps } from './create-fs.types';

import { defaultProps } from './create-fs.defaults';

export function createFS({
  databaseName = defaultProps.databaseName,
  databaseVersion = defaultProps.databaseVersion,
  objectStoreName = defaultProps.objectStoreName,
  rootDirectoryName = defaultProps.rootDirectoryName,
}: CreateFSProps = defaultProps) {
  function initialize() {
    checkIndexedDBSupport();
    validateProps();
  }

  function checkIndexedDBSupport() {
    if (!isIndexedDBSupport()) {
      throw new Error('Your browser does not support indexedDB.');
    }
  }

  function validateProps() {
    validateCreateFSProps({
      databaseName,
      objectStoreName,
      databaseVersion,
      rootDirectoryName,
    });
  }

  const initializeObjectStore = initializeObjectStoreDecorator({
    databaseName,
    databaseVersion,
    objectStoreName,
  });

  const exists = existsDecorator({ rootDirectoryName, initializeObjectStore });

  const isFile = isFileDecorator({
    exists,
    rootDirectoryName,
    initializeObjectStore,
  });

  const remove = removeDecorator({
    exists,
    rootDirectoryName,
    initializeObjectStore,
  });

  const readFile = readFileDecorator({
    isFile,
    rootDirectoryName,
    initializeObjectStore,
  });

  const writeFile = writeFileDecorator({
    exists,
    rootDirectoryName,
    initializeObjectStore,
  });

  const removeFile = removeFileDecorator({
    isFile,
    rootDirectoryName,
    initializeObjectStore,
  });

  const isDirectory = isDirectoryDecorator({
    exists,
    rootDirectoryName,
    initializeObjectStore,
  });

  const readDirectory = readDirectoryDecorator({
    isDirectory,
    rootDirectoryName,
    initializeObjectStore,
  });

  const createDirectory = createDirectoryDecorator({
    exists,
    rootDirectoryName,
    initializeObjectStore,
  });

  const removeDirectory = removeDirectoryDecorator({
    remove,
    isDirectory,
    readDirectory,
    rootDirectoryName,
  });

  initialize();

  return {
    exists,
    isFile,
    remove,
    readFile,
    writeFile,
    removeFile,
    isDirectory,
    readDirectory,
    createDirectory,
    removeDirectory,
    databaseName,
    databaseVersion,
    objectStoreName,
    rootDirectoryName,
  };
}
