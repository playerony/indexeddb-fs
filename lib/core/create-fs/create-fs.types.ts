import { FileEntry, DirectoryEntry } from '@types';
import { ReadDirectoryDecoratorOutput } from './parts';

export type AnyFunction = (...args: any[]) => any;

export interface CreateFsProps {
  databaseName?: string;
  databaseVersion?: number;
  objectStoreName?: string;
  rootDirectoryName?: string;
}

export interface CreateFsOutput {
  databaseName: string;
  databaseVersion: number;
  objectStoreName: string;
  rootDirectoryName: string;
  remove: (fullPath: string) => Promise<void>;
  exists: (fullPath: string) => Promise<boolean>;
  isFile: (fullPath: string) => Promise<boolean>;
  removeFile: (fullPath: string) => Promise<void>;
  isDirectory: (fullPath: string) => Promise<boolean>;
  removeDirectory: (fullPath: string) => Promise<void>;
  readFile: <TData = any>(fullPath: string) => Promise<TData>;
  createDirectory: (fullPath: string) => Promise<DirectoryEntry>;
  directoryDetails: (fullPath: string) => Promise<DirectoryEntry>;
  details: (fullPath: string) => Promise<FileEntry<any> | DirectoryEntry>;
  fileDetails: <TData = any>(fullPath: string) => Promise<FileEntry<TData>>;
  readDirectory: (fullPath: string) => Promise<ReadDirectoryDecoratorOutput>;
  writeFile: <TData = any>(fullPath: string, data: TData) => Promise<FileEntry<TData>>;
}
