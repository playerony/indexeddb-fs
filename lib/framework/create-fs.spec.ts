import { functionImportTest } from '@utils';

import { createFs } from './create-fs.function';

describe('createFs Function', () => {
  functionImportTest(createFs);

  it('should throw an error when the browser does not support indexedDB', () => {
    const copiedIndexedDB = indexedDB;

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    indexedDB = null as any;

    expect(() => createFs()).toThrowErrorMatchingSnapshot();

    // eslint-disable-next-line no-global-assign
    indexedDB = copiedIndexedDB;
  });

  describe('fs object creation', () => {
    it('should throw an error when passed configuration is invalid', () => {
      expect(() =>
        createFs({
          databaseVersion: 1.5,
          rootDirectoryName: '',
          objectStoreName: '  starts with space',
          databaseName: 'too long database name as example of databaseName parameter',
        }),
      ).toThrowErrorMatchingSnapshot();
    });

    it('should return an fs object with the default configuration', () => {
      const { databaseName, databaseVersion, objectStoreName, rootDirectoryName } = createFs();

      expect(databaseVersion).toEqual(1);
      expect(objectStoreName).toEqual('files');
      expect(rootDirectoryName).toEqual('root');
      expect(databaseName).toEqual('indexeddb-fs');
    });
  });

  describe('root directory case', () => {
    it('should create root directory as default', async () => {
      const { isDirectory } = createFs();

      await expect(isDirectory('root')).resolves.toBeTruthy();
    });

    it('user should be able to remove root directory', async () => {
      const { createDirectory, exists, removeDirectory, rootDirectoryName, writeFile } = createFs();

      await createDirectory('files');
      await createDirectory('directories');

      await writeFile('file.txt', 'file');
      await writeFile('files/file.txt', 'file 2');

      await removeDirectory(rootDirectoryName);
      await expect(exists('root')).resolves.toBeTruthy();
      await expect(exists('file.txt')).resolves.toBeFalsy();
      await expect(exists('root/files')).resolves.toBeFalsy();
      await expect(exists('files/file.txt')).resolves.toBeFalsy();
    });

    it('user should be able to read root directory', async () => {
      const { createDirectory, readDirectory, rootDirectoryName, writeFile } = createFs();

      await writeFile('/file.txt', 'content');
      await writeFile('root/file1.txt', 'content');
      await createDirectory('example_directory');

      const { directoriesCount, filesCount, isEmpty } = await readDirectory(rootDirectoryName);

      expect(isEmpty).toBeFalsy();
      expect(filesCount).toEqual(2);
      expect(directoriesCount).toEqual(1);
    });
  });

  it('user should be able to create and remove directory with given name', async () => {
    const {
      createDirectory,
      exists,
      isDirectory,
      isFile,
      readDirectory,
      removeDirectory,
      rootDirectoryName,
      writeFile,
    } = createFs();

    await createDirectory('files');
    await createDirectory('/files/private');
    await createDirectory(`${rootDirectoryName}/files/public`);

    await expect(isDirectory('files')).resolves.toBeTruthy();
    await expect(isDirectory('/files/private')).resolves.toBeTruthy();
    await expect(isDirectory(rootDirectoryName)).resolves.toBeTruthy();
    await expect(isDirectory('root/files/private')).resolves.toBeTruthy();
    await expect(isDirectory(`${rootDirectoryName}/files/private`)).resolves.toBeTruthy();

    await writeFile('files/public/file.txt', 'content');
    await expect(isFile('files/public/file.txt')).resolves.toBeTruthy();
    await expect(exists('files/public/file.txt')).resolves.toBeTruthy();

    await removeDirectory(rootDirectoryName);
    const { directoriesCount, filesCount } = await readDirectory(rootDirectoryName);

    expect(filesCount).toEqual(0);
    expect(directoriesCount).toEqual(0);
  });

  it('user should be able to move, copy and rename files', async () => {
    const {
      copyFile,
      createDirectory,
      exists,
      fileDetails,
      isFile,
      moveFile,
      readDirectory,
      readFile,
      removeDirectory,
      removeFile,
      renameFile,
      rootDirectoryName,
      writeFile,
    } = createFs();

    await createDirectory('files');
    await createDirectory('copied_files');
    await writeFile('files/file.txt', 'content');

    await copyFile('files/file.txt', 'copied_files/copied_file.txt');

    await expect(isFile('files/file.txt')).resolves.toBeTruthy();
    await expect(isFile('copied_files/copied_file.txt')).resolves.toBeTruthy();

    await removeFile('files/file.txt');

    await renameFile('copied_files/copied_file.txt', 'file.txt');
    await expect(exists('copied_files/file.txt')).resolves.toBeTruthy();
    await expect(exists('copied_files/copied_file.txt')).resolves.toBeFalsy();

    await moveFile('copied_files/file.txt', 'files/file.txt');
    await expect(isFile('files/file.txt')).resolves.toBeTruthy();
    await expect(exists('copied_files/file.txt')).resolves.toBeFalsy();

    await moveFile('files/file.txt', 'file.txt');
    await removeDirectory('files');
    await removeDirectory('copied_files');

    const { directoriesCount, filesCount } = await readDirectory(rootDirectoryName);

    expect(filesCount).toEqual(1);
    expect(directoriesCount).toEqual(0);

    const details = await fileDetails('file.txt');

    expect(details.type).toEqual('file');
    expect(details.data).toEqual('content');
    expect(details.name).toEqual('file.txt');
    expect(details.fullPath).toEqual('root/file.txt');
    expect(details.directory).toEqual(rootDirectoryName);

    await expect(readFile('file.txt')).resolves.toEqual('content');
  });
});
