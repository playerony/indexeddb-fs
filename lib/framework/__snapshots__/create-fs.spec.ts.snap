// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createFs Function fs object creation should throw an error when passed configuration is invalid 1`] = `
"Props passed to createFS function are invalid:
[{"path":["databaseName"],"property":"instance.databaseName","message":"does not meet maximum length of 50","schema":{"minLength":4,"maxLength":50,"type":"string","pattern":"^[a-zA-Z0-9_.-]*$"},"instance":"too long database name as example of databaseName parameter","name":"maxLength","argument":50,"stack":"instance.databaseName does not meet maximum length of 50"},{"path":["databaseName"],"property":"instance.databaseName","message":"does not match pattern \\"^[a-zA-Z0-9_.-]*$\\"","schema":{"minLength":4,"maxLength":50,"type":"string","pattern":"^[a-zA-Z0-9_.-]*$"},"instance":"too long database name as example of databaseName parameter","name":"pattern","argument":"^[a-zA-Z0-9_.-]*$","stack":"instance.databaseName does not match pattern \\"^[a-zA-Z0-9_.-]*$\\""},{"path":["objectStoreName"],"property":"instance.objectStoreName","message":"does not match pattern \\"^[a-zA-Z0-9_.-]*$\\"","schema":{"minLength":1,"maxLength":20,"type":"string","pattern":"^[a-zA-Z0-9_.-]*$"},"instance":"  starts with space","name":"pattern","argument":"^[a-zA-Z0-9_.-]*$","stack":"instance.objectStoreName does not match pattern \\"^[a-zA-Z0-9_.-]*$\\""},{"path":["rootDirectoryName"],"property":"instance.rootDirectoryName","message":"does not meet minimum length of 1","schema":{"minLength":1,"maxLength":20,"type":"string","pattern":"^[a-zA-Z0-9_.-]*$"},"instance":"","name":"minLength","argument":1,"stack":"instance.rootDirectoryName does not meet minimum length of 1"},{"path":["databaseVersion"],"property":"instance.databaseVersion","message":"is not of a type(s) integer","schema":{"minimum":1,"maximum":100,"type":"integer"},"instance":1.5,"name":"type","argument":["integer"],"stack":"instance.databaseVersion is not of a type(s) integer"}]"
`;

exports[`createFs Function should throw an error when the browser does not support indexedDB 1`] = `"Your browser does not support indexedDB."`;
